src/core/pointcloud/qgspointcloudindex.h

```diff
@@ -156,29 +157,20 @@ class CORE_EXPORT QgsPointCloudIndex: public QObject
     //! Returns all children of node
     QList<IndexedPointCloudNode> children( const IndexedPointCloudNode &n );
 
-    //! Returns node positions, needs to be scaled and offset applied to get coordinates
-    virtual QVector<qint32> nodePositionDataAsInt32( const IndexedPointCloudNode &n ) = 0;
+    //! Returns all attributes that are stored in the file
+    QgsPointCloudAttributeCollection attributes() const;
 
     /**
-     Loads classes data
-
-    Standard LIDAR classes:
-      0 Created, never classified
-      1 Unclassified
-      2 Ground
-      3 Low Vegetation
-      4 Medium Vegetation
-      5 High Vegetation
-      6 Building
-      7 Low Point (noise)
-      8 Model Key-point (mass point)
-      9 Water
-      10 Reserved for ASPRS Definition
-      11 Reserved for ASPRS Definition
-      12 Overlap Points
-      13-31 Reserved for ASPRS Definition
-    */
-    virtual QVector<char> nodeClassesDataAsChar( const IndexedPointCloudNode &n ) = 0;
+     * Returns node data block
+     *
+     * e.g. positions (needs to be scaled and offset applied to get coordinates) or
+     * classification, intensity or custom attributes
+     *
+     * It is caller responsibility to free the block.
+     *
+     * May return nullptr in case the node is not present or any other problem with loading
+     */
+    virtual QgsPointCloudBlock *nodeData( const IndexedPointCloudNode &n, const QgsPointCloudAttributeCollection &requestedAttributes ) = 0;
```

Comment:
I'm not a fan of passing the attribute collection directly here. I'd rather see this added to a "QgsPointCloudRequest"  (modeled off QgsFeatureRequest) class instead, and the request class used here. That would give us more flexibility to add extra filtering options in future.